### PR TITLE
Added partially extracted slots support for the GroupSlots

### PR DESCRIPTION
**Pull request by NotBioWaste905**
# Description

Added flag `allow_partially_extracted` to the `GroupSlot` class constructor. If `True` non extracted slots from Group Slot would not be overwritten with default values.

# Checklist

- [x] Manual testing
- [x] Autotesting
- [x] Linting
- [x] Tutorials
- [x] Documentation

### CODE REVIEW

File: tests/slots/test_slot_partial_extraction.py
Comment (RLKRo) [inline tests/slots/test_slot_partial_extraction.py:1]: Write unit tests instead.

Add two additional test cases in [`test_slot_types.test_group_slot_extraction`](https://github.com/deeppavlov/chatsky/blob/8306bbbe939bea4c663795a47929894ea8ac9c8b/tests/slots/test_slot_types.py#L103-L124)
(first is the copy of the previous with the option partial extraction allowed; second with partial extraction of nested group slots (e.g. `{"1": {"fail": FailSlot}, "2": {"fail": FailSlot, "success": Slot}}`)
and one test in `test_slot_manager` (to test correct slot storage partial update) also with nested GroupSlot testing.

**Comment by NotBioWaste905** [comment]
llm_response must inherit from BaseResponse

File: tutorials/slots/2_partial_extraction.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            RESPONSE: "Your username is {person.username}. "
+            "Your email is {person.email}.",
+            PRE_RESPONSE: {"fill": proc.FillTemplate()},
+            TRANSITIONS: [Tr(dst=("root", "utter_friend"))],
+        },
+    },
+}
+
+HAPPY_PATH = [
+    ("hi", "Write your username (my username is ...):"),
```
Comment (RLKRo) [inline tutorials/slots/2_partial_extraction.py:171]: Update happy path.

**Review by RLKRo — CHANGES_REQUESTED** [review]

**Comment by RLKRo** [comment]
I don't like name regex in this tutorial.
The last request is very unintuitive ("Mike " with a space at the end).
I think it's better to use other types of data for tutorials that are more easily extracted.

For example, here I use email and bitcoin address which both have fairly decisive regex (i.e. unlikely false positives):
https://github.com/RLKRo/chatsky-ac-tests/blob/main/slot/test_data.json

(regex for these can be found [here](https://github.com/RLKRo/chatsky-ac-tests/blob/main/slot/test_data.json); although email regex is pretty long, [this](https://regexr.com/2ri2c) might be better)

Also, I don't think that `success_only` was really shown in the script and happy path.
It should show how extraction with `success_only=False` looks like (e.g. only pass email for person extraction and it should return `Your username is None. Your email is None.`).

**Comment by NotBioWaste905** [comment]
Consider adding way of defining minimal slots required for the group extraction to succeed
`cnd.GroupSlotExtracted("person",` required=["email"])`

File: chatsky/conditions/slots.py [hunk 2]
```diff
@@ -36,3 +36,31 @@ async def call(self, ctx: Context) -> bool:
...
+
+    def __init__(self, *slots: SlotName, required: List[str] = ["all"]):
+        super().__init__(slots=slots, required=required)
+
+    async def call(self, ctx: Context) -> bool:
+        manager = ctx.framework_data.slot_manager
+        if self.required == ["all"]:
+            return all(manager.is_slot_extracted(slot) for slot in self.slots)
+        else:
+            return all(manager.is_slot_extracted(slot) for slot in self.required)
```
Comment (RLKRo) [inline chatsky/conditions/slots.py:66]: Wait, what is the point of this condition?
Isn't `GroupSlotsExtracted("slot", required=["slot.subslot"])` equivalent to `SlotsExtracted("slot.subslot")`?

Comment (NotBioWaste905) [reply chatsky/conditions/slots.py:66]: Yes indeed. It is supposed to be used like 
`GroupSlotsExtracted(["group_slot_1", "group_slot_2", "group_slot_3"], required=["group_slot_1.subslot_2", "group_slot_2.subslot_3"])`

Possibly can be merged with `SlotsExtracted` in this case?

Comment (RLKRo) [reply chatsky/conditions/slots.py:66]: That's still easier to accomplish with `SlotsExtracted`:
```
SlotsExtracted(["group_slot_1.subslot_2", "group_slot_2.subslot_3", "group_slot_3"])
```

File: tutorials/slots/2_partial_extraction.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            RESPONSE: "Please, send your email and bitcoin address in one message."
+        },
+        "repeat_question": {
+            RESPONSE: "Please, send your bitcoin address and email again."
+        },
+    },
+    "friend_flow": {
+        LOCAL: {
+            PRE_TRANSITION: {
+                "get_slots": proc.Extract("friend", success_only=False)
```
Comment (RLKRo) [inline tutorials/slots/2_partial_extraction.py:137]: No need to use success_only with allow_partial_extraction.

File: tutorials/slots/2_partial_extraction.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+script = {
+    GLOBAL: {
+        TRANSITIONS: [
+            Tr(dst=("user_flow", "ask"), cnd=cnd.Regexp(r"^[sS]tart"))
+        ]
+    },
+    "user_flow": {
+        LOCAL: {
+            PRE_TRANSITION: {
+                "get_slots": proc.Extract("person", success_only=False)
```
Comment (RLKRo) [inline tutorials/slots/2_partial_extraction.py:114]: Showcase success_only=False by first extracting bitcoin only and then extracting email only.
In both cases it should show that only one of the child slots is extracted.

File: tutorials/slots/2_partial_extraction.py [hunk 5]
```diff
@@ -63,150 +61,169 @@
...
+Any slot in this group is saved if and only if that slot is successfully
+extracted.
+
+Group `success_only_extraction` is extracted with the `success_only`
+flag set to True.
+Any slot in this group is saved if and only if all of the slots in the group
+are successfully extracted within a single `Extract` call.
+
+Group `success_only_false` is extracted with the `success_only` set to False.
+Any slot in this group is saved (even if extraction was not successful).
```
Comment (NotBioWaste905) [inline tutorials/slots/2_partial_extraction.py:77]: Every slot in this group is saved?
Not sure, just a grammar thing

Comment (NotBioWaste905) [reply tutorials/slots/2_partial_extraction.py:77]: Overall I like it, it will do

**Review by RLKRo — APPROVED** [review]

**Pull request merged**